Add login and logout endpoints

Jess3Jane opened the pull request:
This should probably not be merged until i document it a little better but as the title says there's now actual endpoints on the api

things to note:
`qaul-http::models` contains the JSON:API models used by the endpoints
authenticator stuff has been moved into `qaul-http::auth`
so much code is required to produce meaningful errors that i've split that out into a different file as it makes the main flow of `login` and `logout` much cleaner

Jess3Jane committed:
allow turning methods back into arrays

Jess3Jane committed:
added `User` and `UserAuth` models

Jess3Jane committed:
added middleware to parse JSON:API documents

Jess3Jane committed:
fix typo in doc comment

Jess3Jane committed:
jsonapi needs to be public to be useful

Jess3Jane committed:
refactor authenticator middleware into own module

Jess3Jane committed:
added a gaurd for jsonapi endpoints

Jess3Jane committed:
add data checks on login

Jess3Jane committed:
update user to use the ~new and improved user data model~

Jess3Jane committed:
authenticator now returns `UserAuth`

Jess3Jane committed:
changed what authenticator puts in map

  - `Authenticator` now puts a copy of itself in the map to allow
     modification of the user map
  - `CurrentUser` allows access to the `UserAuth` of the presently
    logged in user

Jess3Jane committed:
added login endpoint

Jess3Jane committed:
change location of json-api to qaul github

Jess3Jane committed:
made UserAuth's methods public

the enum variants are public anyways so i could (and would)
just reimplement these externally anyways so there's not much
reason for them to be `pub (crate)` instead of `pub`

Jess3Jane committed:
added log out endpoint

Jess3Jane committed:
added login and logout to router

spacekookie commented on libqaul/http-api/src/auth/authenticator.rs:
```diff
@@ -0,0 +1,52 @@
+impl BeforeMiddleware for Authenticator {
+    fn before(&self, req: &mut Request) -> IronResult<()> {
+        if let Some(bearer) = req.headers.get::<Authorization<Bearer>>() {
+            if let Some(identity) = self.tokens.lock().unwrap().get(&bearer.token) {
```
It might be nicer to rewrite this block in a more functional style, i.e. with `.get::<...>().and_then(|bearer| ...).and_then(|identity| ...)`, etc.

Jess3Jane replied:
ah i was having issues getting ownership to work out but double `and_then` worked thankyou

spacekookie commented on libqaul/http-api/src/auth/error.rs:
```diff
@@ -0,0 +1,137 @@
+
+        let detail = match self {
+            AuthError::ConversionError(ObjectConversionError::ImproperType{ expected, got }) => 
+                Some(format!("Primary data should be of type {} but is of type {} instead", 
```
Not something that's really relevant for this PR, but maybe we should keep an eye on is translation services. With this much detail in the error messages, it'd be kinda cool if we could pull in translated strings somehow and let service authors (or us) be able to use some of them to directly display in the presentation layer.

...maybe

Jess3Jane replied:
that'd definitely be a good idea and probably falls under the "configuration management" area of my proposal so I'll definately have a think about it

spacekookie commented on libqaul/http-api/src/auth/login.rs:
```diff
@@ -0,0 +1,104 @@
+    // perform the login
+    let (ident, token) = match qaul.user_login(identity.clone(), &secret) {
+        Ok(QaulUserAuth::Trusted(ident, token)) => (ident, token),
+        Ok(QaulUserAuth::Default(_)) => {
```
Maybe just put the `unreachable!()` on the next line. Maybe we should also rename the `UserAuth` variant here to be more self descriptive that makes the comment obsolete

Jess3Jane replied:
i renamed it to `UserAuth::Untrusted` which hopefully better reflects why that is a meaningless return

spacekookie commented on libqaul/http-api/src/auth/login.rs:
```diff
@@ -0,0 +1,104 @@
+
+    // return the grant
+    // so you'd think the id here would be the user id right?
+    // NO
```
Oof

Jess3Jane replied:
yeah... in  this case i might make it just the token type and return a success message for cookie grants but in general this is a problem and this is roughly the community standard solution

which is obviously still kinda weird but >.>

spacekookie reviewed (commented):
Hey, this is looking pretty good!
I haven't gotten all the way through the PR yet, but I left you some notes. I'll try to get the rest of it later. 

Also, could you write a general short overview of what's still missing from this PR?

Jess3Jane commented:
The major thing i need to work on is documenting this
there's a bit of the authenticator stuff that will be needed by any services wishing to present an http api and those should get doc comments
i also need to document endpoints and entities but i'm not really sure where to do that?
finally i'd like to add another grant type to allow for granting server side cookies instead of returning a token

Jess3Jane committed:
rewrite authenticator to use functional style

Jess3Jane committed:
rename `UserAuth::Default` to `UserAuth::Untrusted`

this better reflects its meaning

Jess3Jane committed:
add grant type, change success message generation

  - you can now request a token grant or a (currently non-functional)
cookie grant
  - token grants now have the token in the id instead of as an attribute
  - the success model now has a method to construct a new success
message with the current time as an id

Jess3Jane committed:
bump versions

Jess3Jane committed:
mark cookie grant as unimplemented

Jess3Jane commented:
added the grant type but login cookie will be a separate pr later

Jess3Jane committed:
document and update visiblity

Jess3Jane committed:
added http-api documentation book

Jess3Jane committed:
Note existance of http-api book

Jess3Jane commented:
:heavy_check_mark: document public structures
:question: i have documented the http-api in its own book at the moment but i'm not sure that's the best place for it

spacekookie commented on docs/contributers/src/SUMMARY.md:
```diff
@@ -9,10 +9,12 @@
     - [GPL License (Version 3)](./legal/gpl-3.0.md)
-    - [Creative Commons Attribution 4.0 International License](./legal/cc-by.md)
\ No newline at end of file
+    - [Creative Commons Attribution 4.0 International License](./legal/cc-by.md)
```
What happened with that line? :sweat_smile:

Jess3Jane replied:
?-? i haven't a clue

spacekookie commented on docs/http-api/src/entities/user.md:
```diff
@@ -0,0 +1,19 @@
+
+## Attributes
+### `display_name: String` _(optional)_
+A human readable display-name (like @foobar)
```
Maybe put "@foobar" into `code` blocks

spacekookie commented on docs/http-api/src/topics/authentication.md:
```diff
@@ -0,0 +1,45 @@
+			"grant_type": "token",
+		}
+	}
+}
```
This code block seems to be indented with a mix of tabs and spaces

Jess3Jane replied:
That's very strange as it's not on my local copy

spacekookie reviewed (commented):
I skimmed through the documentation you added and it looks good at a glance. I made a few comments about formatting, but those are all pretty minor.

Also, I kinda feel like putting the `http` docs in the `contributers` guide is gonna be nicer. It means people will only need to get a single book, it'll be easier to host, and also it just adds more context to everything people are looking at.

Not sure if it's possible to make `mdbook` fold away items until they're being expanded, but even if not, I think putting it all in the same book might still be nicer

Jess3Jane committed:
Failed login should error

previously when login failed it just didn't set the user field
however, it is more correct for it to error the application

if no login information is provided it will still pass

Jess3Jane committed:
put @foobar in code blocks

Jess3Jane committed:
update authentication with new errors

Jess3Jane commented:
Unfortunately there does not currently seem to be a way to fold interior chapters, although i agree that would be ideal
https://github.com/rust-lang-nursery/mdBook/issues/187

Jess3Jane commented:
i have reformatted the authenticator again because it should probably not continue as if nothing is wrong when a request has an invalid token

with that other than moving the documentation which i'm not sure if we want to do i think that's all i have for this pr so i'll mark it as ready

spacekookie reviewed (approved):
I'll just merge this as is for now. We can yakshave the documentation issue later. I think having it all i a single book would be good. Maybe we can patch mdbook to make that happen. But that's beyond the scope of this PR.

Otherwise this looks really good and thank you for all your effort! :heart:
